tlscommon.c

```diff
@@ -626,13 +628,199 @@ static int certattr_matchip(GENERAL_NAME *gn, struct certattrmatch *match){
         && !memcmp(ASN1_STRING_get0_data(gn->d.iPAddress), &match->ipaddr, l)) ? 1 : 0 ;
 }
 
+static char* concatenate(char* s1, char* s2) {
+    if(s1 == NULL) {
+        return s2;
+    } 
+    if(s2 == NULL) {
+        return s1;
+    }
+    if(s1 == NULL && s2 == NULL) {
+        return NULL;
+    }
+    char* result = (char*) calloc(strlen(s1) + strlen(s2), sizeof(char)); 
+    if(result == NULL) {
+        return NULL;
+    }
+    memcpy(result, s1, strlen(s1));
+    memcpy(result + strlen(s1), s2, strlen(s2));
+    return result;
+}
+
+static char* append(int n, ...) {
+   char* val = NULL;
+   va_list ap;
+   int i;
+
+   va_start(ap, n);
+   char* result = NULL;
+   for(i = 0; i < n; i++) {
+      val = va_arg(ap, char*);
+      result = concatenate(result, val);
+   }
+   va_end(ap);
+   return result;
+}
+
+static char* extractFirstPortion(char* dns, char c) {
+    char* result = NULL;
+    char *firstPos = strchr(dns, c);
+    if(firstPos == NULL) {
+        return result;
+    }
+    int lenOfFirstPortion = strlen(dns) - strlen(firstPos);
+    result = (char*) calloc(lenOfFirstPortion, sizeof(char));
+    if(result == NULL) {
+        return NULL;
+    }
+    memcpy(result, dns, lenOfFirstPortion);
+    return result;
+}
+
+static int regexMatcher(char* pattern, char* str) {
+    regex_t    regex;
+    char buffer[100];
+    
+    int reti = regcomp(&regex, pattern, REG_EXTENDED);
+    if (reti) {
+        regerror(reti, &regex, buffer, 100);
+        debug(DBG_ERR, "regcomp failed with '%s'\n", buffer);
+        return 0;
+    }        
+    reti = regexec(&regex, str, 0, NULL, 0);
+    if(reti) {             
+        return 0;                                            
+    }
+    return 1;
+}
+
+static char* calculateLenOfLabel(int len) {
+    char str[3];
+    sprintf(str, "%d", len);
```

Comment:
Buffer overflow here.  len can (in general) be the size of an integer.  So you need more than 3 bytes to store a 32-bit decimal value.